// @react-three/drei/core/Segments.js
import _extends from '@babel/runtime/helpers/esm/extends';
import * as THREE from 'three';
import * as React from 'react';
import { useFrame } from '@react-three/fiber';
import { Line2, LineMaterial, LineSegmentsGeometry } from 'three-stdlib';

const context = /*#__PURE__*/React.createContext(null);

const arrColor = color => color instanceof THREE.Color ? color.toArray() : color;

const arrPos = pos => pos instanceof THREE.Vector3 ? pos.toArray() : pos;

const Segments = /*#__PURE__*/React.forwardRef((props, forwardedRef) => {
  const {
    limit = 1000,
    lineWidth = 1.0,
    children,
    ...rest
  } = props;
  const [segments, setSegments] = React.useState([]);
  const [line] = React.useState(() => new Line2());
  const [material] = React.useState(() => new LineMaterial());
  const [geometry] = React.useState(() => new LineSegmentsGeometry());
  const [resolution] = React.useState(() => new THREE.Vector2(512, 512));
  const [positions] = React.useState(() => Array(limit).fill(0));
  const [colors] = React.useState(() => Array(limit).fill(1));
  const api = React.useMemo(() => ({
    subscribe: ref => {
      setSegments(segments => [...segments, ref]);
      return () => setSegments(segments => segments.filter(item => item.current !== ref.current));
    }
  }), []);
  useFrame(() => {
    positions.length = segments.length * 6;
    colors.length = segments.length * 6;
    segments.forEach((segmentRef, i) => {
      const segment = segmentRef.current;

      if (segment) {
        const segmentStart = arrPos(segment.start);
        const segmentEnd = arrPos(segment.end);
        const segmentColor = arrColor(segment.color);

        for (var j = 0; j < 3; j++) {
          positions[i * 6 + j] = segmentStart[j];
          positions[i * 6 + j + 3] = segmentEnd[j];
          colors[i * 6 + j] = segmentColor[j];
          colors[i * 6 + j + 3] = segmentColor[j];
        }
      }
    });
    geometry.setColors(colors);
    geometry.setPositions(positions);
    line.computeLineDistances();
  });
  return /*#__PURE__*/React.createElement("primitive", {
    object: line,
    ref: forwardedRef
  }, /*#__PURE__*/React.createElement("primitive", {
    object: geometry,
    attach: "geometry"
  }), /*#__PURE__*/React.createElement("primitive", _extends({
    object: material,
    attach: "material",
    vertexColors: true,
    resolution: resolution,
    linewidth: lineWidth
  }, rest)), /*#__PURE__*/React.createElement(context.Provider, {
    value: api
  }, children));
});
const Segment = /*#__PURE__*/React.forwardRef((props, forwardedRef) => {
  const {
    start,
    end,
    color
  } = props;
  const api = React.useContext(context);
  if (!api) throw 'Segment must used inside Segments component.';
  const ref = React.useRef({
    start: [0, 0, 0],
    end: [0, 0, 0],
    color: [1, 1, 1]
  });
  React.useLayoutEffect(() => api.subscribe(forwardedRef || ref), []);
  React.useLayoutEffect(() => {
    var currRef = forwardedRef || ref;
    start && Object.assign(currRef.current, {
      start
    });
    end && Object.assign(currRef.current, {
      end
    });
    color && Object.assign(currRef.current, {
      color
    });
  }, [start, end, color]);
  return null;
});

export { Segment, Segments };
